<a href="https://colab.research.google.com/github/kur114/notebook-toys/blob/main/hyperparameter_tuning_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
# For tips on running notebooks in Google Colab, see
# https://pytorch.org/tutorials/beginner/colab
%matplotlib inline

!pip install ray[tune] torchvision

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.7/101.7 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.3/66.3 MB 10.6 MB/s eta 0:00:00


Hyperparameter tuning with Ray Tune
===================================

Hyperparameter tuning can make the difference between an average model
and a highly accurate one. Often simple things like choosing a different
learning rate or changing a network layer size can have a dramatic
impact on your model performance.

Fortunately, there are tools that help with finding the best combination
of parameters. [Ray Tune](https://docs.ray.io/en/latest/tune.html) is an
industry standard tool for distributed hyperparameter tuning. Ray Tune
includes the latest hyperparameter search algorithms, integrates with
various analysis libraries, and natively supports distributed training
through [Ray\'s distributed machine learning engine](https://ray.io/).

In this tutorial, we will show you how to integrate Ray Tune into your
PyTorch training workflow. We will extend [this tutorial from the
PyTorch
documentation](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html)
for training a CIFAR10 image classifier.

As you will see, we only need to add some slight modifications. In
particular, we need to

1.  wrap data loading and training in functions,
2.  make some network parameters configurable,
3.  add checkpointing (optional),
4.  and define the search space for the model tuning

|

To run this tutorial, please make sure the following packages are
installed:

-   `ray[tune]`: Distributed hyperparameter tuning library
-   `torchvision`: For the data transformers

Setup / Imports
---------------

Let\'s start with the imports:


In [5]:
from functools import partial
import os
import tempfile
from pathlib import Path
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import random_split
import torchvision
import torchvision.transforms as transforms
from ray import tune
from ray import train
from ray.train import Checkpoint, get_checkpoint
from ray.tune.schedulers import ASHAScheduler
import ray.cloudpickle as pickle

Most of the imports are needed for building the PyTorch model. Only the
last imports are for Ray Tune.

Data loaders
============

We wrap the data loaders in their own function and pass a global data
directory. This way we can share a data directory between different
trials.


In [6]:
def load_data(data_dir="./data"):
    transform = transforms.Compose(
        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
    )

    trainset = torchvision.datasets.CIFAR10(
        root=data_dir, train=True, download=True, transform=transform
    )

    testset = torchvision.datasets.CIFAR10(
        root=data_dir, train=False, download=True, transform=transform
    )

    return trainset, testset

Configurable neural network
===========================

We can only tune those parameters that are configurable. In this
example, we can specify the layer sizes of the fully connected layers:


In [7]:
class Net(nn.Module):
    def __init__(self, l1=120, l2=84):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, l1)
        self.fc2 = nn.Linear(l1, l2)
        self.fc3 = nn.Linear(l2, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = torch.flatten(x, 1)  # flatten all dimensions except batch
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

The train function
==================

Now it gets interesting, because we introduce some changes to the
example [from the PyTorch
documentation](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html).

We wrap the training script in a function
`train_cifar(config, data_dir=None)`. The `config` parameter will
receive the hyperparameters we would like to train with. The `data_dir`
specifies the directory where we load and store the data, so that
multiple runs can share the same data source. We also load the model and
optimizer state at the start of the run, if a checkpoint is provided.
Further down in this tutorial you will find information on how to save
the checkpoint and what it is used for.

``` {.python}
net = Net(config["l1"], config["l2"])

checkpoint = get_checkpoint()
if checkpoint:
    with checkpoint.as_directory() as checkpoint_dir:
        data_path = Path(checkpoint_dir) / "data.pkl"
        with open(data_path, "rb") as fp:
            checkpoint_state = pickle.load(fp)
        start_epoch = checkpoint_state["epoch"]
        net.load_state_dict(checkpoint_state["net_state_dict"])
        optimizer.load_state_dict(checkpoint_state["optimizer_state_dict"])
else:
    start_epoch = 0
```

The learning rate of the optimizer is made configurable, too:

``` {.python}
optimizer = optim.SGD(net.parameters(), lr=config["lr"], momentum=0.9)
```

We also split the training data into a training and validation subset.
We thus train on 80% of the data and calculate the validation loss on
the remaining 20%. The batch sizes with which we iterate through the
training and test sets are configurable as well.

Adding (multi) GPU support with DataParallel
--------------------------------------------

Image classification benefits largely from GPUs. Luckily, we can
continue to use PyTorch\'s abstractions in Ray Tune. Thus, we can wrap
our model in `nn.DataParallel` to support data parallel training on
multiple GPUs:

``` {.python}
device = "cpu"
if torch.cuda.is_available():
    device = "cuda:0"
    if torch.cuda.device_count() > 1:
        net = nn.DataParallel(net)
net.to(device)
```

By using a `device` variable we make sure that training also works when
we have no GPUs available. PyTorch requires us to send our data to the
GPU memory explicitly, like this:

``` {.python}
for i, data in enumerate(trainloader, 0):
    inputs, labels = data
    inputs, labels = inputs.to(device), labels.to(device)
```

The code now supports training on CPUs, on a single GPU, and on multiple
GPUs. Notably, Ray also supports [fractional
GPUs](https://docs.ray.io/en/master/using-ray-with-gpus.html#fractional-gpus)
so we can share GPUs among trials, as long as the model still fits on
the GPU memory. We\'ll come back to that later.

Communicating with Ray Tune
---------------------------

The most interesting part is the communication with Ray Tune:

``` {.python}
checkpoint_data = {
    "epoch": epoch,
    "net_state_dict": net.state_dict(),
    "optimizer_state_dict": optimizer.state_dict(),
}
with tempfile.TemporaryDirectory() as checkpoint_dir:
    data_path = Path(checkpoint_dir) / "data.pkl"
    with open(data_path, "wb") as fp:
        pickle.dump(checkpoint_data, fp)

    checkpoint = Checkpoint.from_directory(checkpoint_dir)
    train.report(
        {"loss": val_loss / val_steps, "accuracy": correct / total},
        checkpoint=checkpoint,
    )
```

Here we first save a checkpoint and then report some metrics back to Ray
Tune. Specifically, we send the validation loss and accuracy back to Ray
Tune. Ray Tune can then use these metrics to decide which hyperparameter
configuration lead to the best results. These metrics can also be used
to stop bad performing trials early in order to avoid wasting resources
on those trials.

The checkpoint saving is optional, however, it is necessary if we wanted
to use advanced schedulers like [Population Based
Training](https://docs.ray.io/en/latest/tune/examples/pbt_guide.html).
Also, by saving the checkpoint we can later load the trained models and
validate them on a test set. Lastly, saving checkpoints is useful for
fault tolerance, and it allows us to interrupt training and continue
training later.

Full training function
----------------------

The full code example looks like this:


In [8]:
def train_cifar(config, data_dir=None):
    net = Net(config["l1"], config["l2"])

    device = "cpu"
    if torch.cuda.is_available():
        device = "cuda:0"
        if torch.cuda.device_count() > 1:
            net = nn.DataParallel(net)
    net.to(device)

    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(net.parameters(), lr=config["lr"], momentum=0.9)

    checkpoint = get_checkpoint()
    if checkpoint:
        with checkpoint.as_directory() as checkpoint_dir:
            data_path = Path(checkpoint_dir) / "data.pkl"
            with open(data_path, "rb") as fp:
                checkpoint_state = pickle.load(fp)
            start_epoch = checkpoint_state["epoch"]
            net.load_state_dict(checkpoint_state["net_state_dict"])
            optimizer.load_state_dict(checkpoint_state["optimizer_state_dict"])
    else:
        start_epoch = 0

    trainset, testset = load_data(data_dir)

    test_abs = int(len(trainset) * 0.8)
    train_subset, val_subset = random_split(
        trainset, [test_abs, len(trainset) - test_abs]
    )

    trainloader = torch.utils.data.DataLoader(
        train_subset, batch_size=int(config["batch_size"]), shuffle=True, num_workers=8
    )
    valloader = torch.utils.data.DataLoader(
        val_subset, batch_size=int(config["batch_size"]), shuffle=True, num_workers=8
    )

    for epoch in range(start_epoch, 10):  # loop over the dataset multiple times
        running_loss = 0.0
        epoch_steps = 0
        for i, data in enumerate(trainloader, 0):
            # get the inputs; data is a list of [inputs, labels]
            inputs, labels = data
            inputs, labels = inputs.to(device), labels.to(device)

            # zero the parameter gradients
            optimizer.zero_grad()

            # forward + backward + optimize
            outputs = net(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            # print statistics
            running_loss += loss.item()
            epoch_steps += 1
            if i % 2000 == 1999:  # print every 2000 mini-batches
                print(
                    "[%d, %5d] loss: %.3f"
                    % (epoch + 1, i + 1, running_loss / epoch_steps)
                )
                running_loss = 0.0

        # Validation loss
        val_loss = 0.0
        val_steps = 0
        total = 0
        correct = 0
        for i, data in enumerate(valloader, 0):
            with torch.no_grad():
                inputs, labels = data
                inputs, labels = inputs.to(device), labels.to(device)

                outputs = net(inputs)
                _, predicted = torch.max(outputs.data, 1)
                total += labels.size(0)
                correct += (predicted == labels).sum().item()

                loss = criterion(outputs, labels)
                val_loss += loss.cpu().numpy()
                val_steps += 1

        checkpoint_data = {
            "epoch": epoch,
            "net_state_dict": net.state_dict(),
            "optimizer_state_dict": optimizer.state_dict(),
        }
        with tempfile.TemporaryDirectory() as checkpoint_dir:
            data_path = Path(checkpoint_dir) / "data.pkl"
            with open(data_path, "wb") as fp:
                pickle.dump(checkpoint_data, fp)

            checkpoint = Checkpoint.from_directory(checkpoint_dir)
            train.report(
                {"loss": val_loss / val_steps, "accuracy": correct / total},
                checkpoint=checkpoint,
            )

    print("Finished Training")

As you can see, most of the code is adapted directly from the original
example.

Test set accuracy
=================

Commonly the performance of a machine learning model is tested on a
hold-out test set with data that has not been used for training the
model. We also wrap this in a function:


In [10]:
def test_accuracy(net, device="cpu"):
    trainset, testset = load_data()

    testloader = torch.utils.data.DataLoader(
        testset, batch_size=4, shuffle=False, num_workers=2
    )

    correct = 0
    total = 0
    with torch.no_grad():
        for data in testloader:
            images, labels = data
            images, labels = images.to(device), labels.to(device)
            outputs = net(images)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    return correct / total

The function also expects a `device` parameter, so we can do the test
set validation on a GPU.

Configuring the search space
============================

Lastly, we need to define Ray Tune\'s search space. Here is an example:

``` {.python}
config = {
    "l1": tune.choice([2 ** i for i in range(9)]),
    "l2": tune.choice([2 ** i for i in range(9)]),
    "lr": tune.loguniform(1e-4, 1e-1),
    "batch_size": tune.choice([2, 4, 8, 16])
}
```

The `tune.choice()` accepts a list of values that are uniformly sampled
from. In this example, the `l1` and `l2` parameters should be powers of
2 between 4 and 256, so either 4, 8, 16, 32, 64, 128, or 256. The `lr`
(learning rate) should be uniformly sampled between 0.0001 and 0.1.
Lastly, the batch size is a choice between 2, 4, 8, and 16.

At each trial, Ray Tune will now randomly sample a combination of
parameters from these search spaces. It will then train a number of
models in parallel and find the best performing one among these. We also
use the `ASHAScheduler` which will terminate bad performing trials
early.

We wrap the `train_cifar` function with `functools.partial` to set the
constant `data_dir` parameter. We can also tell Ray Tune what resources
should be available for each trial:

``` {.python}
gpus_per_trial = 2
# ...
result = tune.run(
    partial(train_cifar, data_dir=data_dir),
    resources_per_trial={"cpu": 8, "gpu": gpus_per_trial},
    config=config,
    num_samples=num_samples,
    scheduler=scheduler,
    checkpoint_at_end=True)
```

You can specify the number of CPUs, which are then available e.g. to
increase the `num_workers` of the PyTorch `DataLoader` instances. The
selected number of GPUs are made visible to PyTorch in each trial.
Trials do not have access to GPUs that haven\'t been requested for them
- so you don\'t have to care about two trials using the same set of
resources.

Here we can also specify fractional GPUs, so something like
`gpus_per_trial=0.5` is completely valid. The trials will then share
GPUs among each other. You just have to make sure that the models still
fit in the GPU memory.

After training the models, we will find the best performing one and load
the trained network from the checkpoint file. We then obtain the test
set accuracy and report everything by printing.

The full main function looks like this:


In [11]:
def main(num_samples=10, max_num_epochs=10, gpus_per_trial=2):
    data_dir = os.path.abspath("./data")
    load_data(data_dir)
    config = {
        "l1": tune.choice([2**i for i in range(9)]),
        "l2": tune.choice([2**i for i in range(9)]),
        "lr": tune.loguniform(1e-4, 1e-1),
        "batch_size": tune.choice([2, 4, 8, 16]),
    }
    scheduler = ASHAScheduler(
        metric="loss",
        mode="min",
        max_t=max_num_epochs,
        grace_period=1,
        reduction_factor=2,
    )
    result = tune.run(
        partial(train_cifar, data_dir=data_dir),
        resources_per_trial={"cpu": 2, "gpu": gpus_per_trial},
        config=config,
        num_samples=num_samples,
        scheduler=scheduler,
    )

    best_trial = result.get_best_trial("loss", "min", "last")
    print(f"Best trial config: {best_trial.config}")
    print(f"Best trial final validation loss: {best_trial.last_result['loss']}")
    print(f"Best trial final validation accuracy: {best_trial.last_result['accuracy']}")

    best_trained_model = Net(best_trial.config["l1"], best_trial.config["l2"])
    device = "cpu"
    if torch.cuda.is_available():
        device = "cuda:0"
        if gpus_per_trial > 1:
            best_trained_model = nn.DataParallel(best_trained_model)
    best_trained_model.to(device)

    best_checkpoint = result.get_best_checkpoint(trial=best_trial, metric="accuracy", mode="max")
    with best_checkpoint.as_directory() as checkpoint_dir:
        data_path = Path(checkpoint_dir) / "data.pkl"
        with open(data_path, "rb") as fp:
            best_checkpoint_data = pickle.load(fp)

        best_trained_model.load_state_dict(best_checkpoint_data["net_state_dict"])
        test_acc = test_accuracy(best_trained_model, device)
        print("Best trial test set accuracy: {}".format(test_acc))


if __name__ == "__main__":
    # You can change the number of GPUs per trial here:
    main(num_samples=10, max_num_epochs=10, gpus_per_trial=0)

100%|██████████| 170M/170M [00:04<00:00, 36.2MB/s]


Extracting /content/data/cifar-10-python.tar.gz to /content/data
Files already downloaded and verified


2024-11-14 07:47:25,672	INFO worker.py:1819 -- Started a local Ray instance.
2024-11-14 07:47:27,944	INFO tune.py:253 -- Initializing Ray automatically. For cluster usage or custom Ray initialization, call `ray.init(...)` before `tune.run(...)`.


+--------------------------------------------------------------------+
| Configuration for experiment     train_cifar_2024-11-14_07-47-28   |
+--------------------------------------------------------------------+
| Search algorithm                 BasicVariantGenerator             |
| Scheduler                        AsyncHyperBandScheduler           |
| Number of trials                 10                                |
+--------------------------------------------------------------------+

View detailed results here: /root/ray_results/train_cifar_2024-11-14_07-47-28
To visualize your results with TensorBoard, run: `tensorboard --logdir /tmp/ray/session_2024-11-14_07-47-22_827631_657/artifacts/2024-11-14_07-47-28/train_cifar_2024-11-14_07-47-28/driver_artifacts`

Trial status: 10 PENDING
Current time: 2024-11-14 07:47:28. Total running time: 0s
Logical resource usage: 0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+-------------------------------------------------------------------

(func pid=4398) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=4398)   warnings.warn(


(func pid=4398) [1,  2000] loss: 2.303

Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:47:58. Total running time: 30s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+-------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size |
+-------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8 |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4 |
| train_cifar_b2154_00003   PENDING       4      8   0.0417593                8 |
| train_cifar_b2154_00004   PENDING      16     64   0.028188                16 |
| train_cifar_b2154_00005   PENDING       1     16   0.000271213              2 |
| train_cifar_b2154_00006   PENDING     256    128  

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000000)


(func pid=4398) [2,  2000] loss: 2.185

Trial train_cifar_b2154_00000 finished iteration 2 at 2024-11-14 07:48:28. Total running time: 1min 0s
+------------------------------------------------------------+
| Trial train_cifar_b2154_00000 result                       |
+------------------------------------------------------------+
| checkpoint_dir_name                      checkpoint_000001 |
| time_this_iter_s                                  26.34421 |
| time_total_s                                      55.06398 |
| training_iteration                                       2 |
| accuracy                                            0.2613 |
| loss                                               2.00415 |
+------------------------------------------------------------+
Trial train_cifar_b2154_00000 saved a checkpoint for iteration 2 at: (local)/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000001


(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000001)



Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:48:28. Total running time: 1min 0s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        2             55.064   2.00415       0.2613 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                                    |
| train_cifar_b2154_00003  

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000002)



Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:48:58. Total running time: 1min 30s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        3            81.7606   1.79185       0.3496 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                                    |
| train_cifar_b2154_00003 

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000003)



Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:49:28. Total running time: 2min 0s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        4            108.288   1.62043       0.4061 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                                    |
| train_cifar_b2154_00003  

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000004)



Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:49:58. Total running time: 2min 30s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        5            134.085   1.54911       0.4326 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                                    |
| train_cifar_b2154_00003 

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000005)



Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:50:28. Total running time: 3min 0s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        6            162.317   1.51073       0.4471 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                                    |
| train_cifar_b2154_00003  

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000006)



Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:50:58. Total running time: 3min 30s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        7            188.338   1.45022       0.4795 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                                    |
| train_cifar_b2154_00003 

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000007)


(func pid=4398) [9,  2000] loss: 1.362

Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:51:28. Total running time: 4min 0s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        8            214.915   1.39333       0.4988 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                          

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000008)


(func pid=4398) [10,  2000] loss: 1.321

Trial status: 1 RUNNING | 9 PENDING
Current time: 2024-11-14 07:51:58. Total running time: 4min 30s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+----------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status       l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+----------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   RUNNING     128     32   0.000342318             16        9            241.381   1.37772       0.5033 |
| train_cifar_b2154_00001   PENDING      32    128   0.000232848              8                                                    |
| train_cifar_b2154_00002   PENDING       8      4   0.00603131               4                                        

(func pid=4398) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00000_0_batch_size=16,l1=128,l2=32,lr=0.0003_2024-11-14_07-47-28/checkpoint_000009)



Trial train_cifar_b2154_00001 started with configuration:
+--------------------------------------------------+
| Trial train_cifar_b2154_00001 config             |
+--------------------------------------------------+
| batch_size                                     8 |
| l1                                            32 |
| l2                                           128 |
| lr                                       0.00023 |
+--------------------------------------------------+
(func pid=6458) Files already downloaded and verified
(func pid=6458) Files already downloaded and verified


(func pid=6458) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=6458)   warnings.warn(


(func pid=6458) [1,  2000] loss: 2.303

Trial status: 1 TERMINATED | 1 RUNNING | 8 PENDING
Current time: 2024-11-14 07:52:29. Total running time: 5min 0s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00001   RUNNING        32    128   0.000232848              8                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10            268.076   1.32249       0.5215 |
| train_cifar_b2154_00002   PENDING         8      4   0.00603131               4               

(func pid=6458) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00001_1_batch_size=8,l1=32,l2=128,lr=0.0002_2024-11-14_07-47-28/checkpoint_000000)



Trial train_cifar_b2154_00002 started with configuration:
+--------------------------------------------------+
| Trial train_cifar_b2154_00002 config             |
+--------------------------------------------------+
| batch_size                                     4 |
| l1                                             8 |
| l2                                             4 |
| lr                                       0.00603 |
+--------------------------------------------------+
(func pid=6799) Files already downloaded and verified
(func pid=6799) Files already downloaded and verified


(func pid=6799) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=6799)   warnings.warn(



Trial status: 2 TERMINATED | 1 RUNNING | 7 PENDING
Current time: 2024-11-14 07:52:59. Total running time: 5min 31s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00002   RUNNING         8      4   0.00603131               4                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |

(func pid=6799) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00002_2_batch_size=4,l1=8,l2=4,lr=0.0060_2024-11-14_07-47-28/checkpoint_000000)



Trial train_cifar_b2154_00003 started with configuration:
+--------------------------------------------------+
| Trial train_cifar_b2154_00003 config             |
+--------------------------------------------------+
| batch_size                                     8 |
| l1                                             4 |
| l2                                             8 |
| lr                                       0.04176 |
+--------------------------------------------------+
(func pid=7231) Files already downloaded and verified
(func pid=7231) Files already downloaded and verified


(func pid=7231) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=7231)   warnings.warn(



Trial status: 3 TERMINATED | 1 RUNNING | 6 PENDING
Current time: 2024-11-14 07:53:59. Total running time: 6min 31s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00003   RUNNING         4      8   0.0417593                8                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |

(func pid=7231) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00003_3_batch_size=8,l1=4,l2=8,lr=0.0418_2024-11-14_07-47-28/checkpoint_000000)



Trial train_cifar_b2154_00004 started with configuration:
+--------------------------------------------------+
| Trial train_cifar_b2154_00004 config             |
+--------------------------------------------------+
| batch_size                                    16 |
| l1                                            16 |
| l2                                            64 |
| lr                                       0.02819 |
+--------------------------------------------------+
(func pid=7574) Files already downloaded and verified
(func pid=7574) Files already downloaded and verified


(func pid=7574) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=7574)   warnings.warn(



Trial status: 4 TERMINATED | 1 RUNNING | 5 PENDING
Current time: 2024-11-14 07:54:59. Total running time: 7min 31s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00004   RUNNING        16     64   0.028188                16                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |

(func pid=7574) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00004_4_batch_size=16,l1=16,l2=64,lr=0.0282_2024-11-14_07-47-28/checkpoint_000000)


(func pid=7574) [2,  2000] loss: 2.003

Trial status: 4 TERMINATED | 1 RUNNING | 5 PENDING
Current time: 2024-11-14 07:55:29. Total running time: 8min 1s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00004   RUNNING        16     64   0.028188                16        1            27.8074   2.00314       0.2295 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1      

(func pid=7574) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00004_4_batch_size=16,l1=16,l2=64,lr=0.0282_2024-11-14_07-47-28/checkpoint_000001)



Trial train_cifar_b2154_00005 started with configuration:
+--------------------------------------------------+
| Trial train_cifar_b2154_00005 config             |
+--------------------------------------------------+
| batch_size                                     2 |
| l1                                             1 |
| l2                                            16 |
| lr                                       0.00027 |
+--------------------------------------------------+
(func pid=8067) Files already downloaded and verified
(func pid=8067) Files already downloaded and verified


(func pid=8067) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=8067)   warnings.warn(


(func pid=8067) [1,  2000] loss: 2.311
(func pid=8067) [1,  4000] loss: 1.153

Trial status: 5 TERMINATED | 1 RUNNING | 4 PENDING
Current time: 2024-11-14 07:55:59. Total running time: 8min 31s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00005   RUNNING         1     16   0.000271213              2                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0

(func pid=8067) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00005_5_batch_size=2,l1=1,l2=16,lr=0.0003_2024-11-14_07-47-28/checkpoint_000000)



Trial train_cifar_b2154_00006 started with configuration:
+--------------------------------------------------+
| Trial train_cifar_b2154_00006 config             |
+--------------------------------------------------+
| batch_size                                     2 |
| l1                                           256 |
| l2                                           128 |
| lr                                       0.00632 |
+--------------------------------------------------+
(func pid=8669) Files already downloaded and verified
(func pid=8669) Files already downloaded and verified


(func pid=8669) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=8669)   warnings.warn(



Trial status: 6 TERMINATED | 1 RUNNING | 3 PENDING
Current time: 2024-11-14 07:57:29. Total running time: 10min 1s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00006   RUNNING       256    128   0.00631651               2                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |

(func pid=8669) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00006_6_batch_size=2,l1=256,l2=128,lr=0.0063_2024-11-14_07-47-28/checkpoint_000000)



Trial status: 6 TERMINATED | 1 RUNNING | 3 PENDING
Current time: 2024-11-14 07:59:29. Total running time: 12min 1s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00006   RUNNING       256    128   0.00631651               2        1           116.267    2.0438        0.245  |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |

(func pid=8669) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00006_6_batch_size=2,l1=256,l2=128,lr=0.0063_2024-11-14_07-47-28/checkpoint_000001)



Trial status: 7 TERMINATED | 3 PENDING
Current time: 2024-11-14 08:01:30. Total running time: 14min 1s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |
| train_cifar_b2154_00002   TERMINATED      8      4   0.00603131               4        1            58.1518   2.29459       0.1152 |
| train_cif

(func pid=9959) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=9959)   warnings.warn(


(func pid=9959) [1,  2000] loss: 2.310
(func pid=9959) [1,  4000] loss: 1.148

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:02:00. Total running time: 14min 32s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000000)


(func pid=9959) [2,  2000] loss: 1.873
(func pid=9959) [2,  4000] loss: 0.907

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:03:00. Total running time: 15min 32s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        1            56.9495   1.90943       0.2662 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000001)



Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:03:30. Total running time: 16min 2s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        2           113.686    1.71001       0.3468 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000002)



Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:04:30. Total running time: 17min 2s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        3           169.386    1.65066       0.3514 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000003)


(func pid=9959) [5,  2000] loss: 1.489

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:05:30. Total running time: 18min 2s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        4           225.396    1.56396       0.3995 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1     

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000004)


(func pid=9959) [6,  2000] loss: 1.407

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:06:30. Total running time: 19min 2s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        5           281.388    1.41273       0.4706 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1     

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000005)


(func pid=9959) [7,  2000] loss: 1.350

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:07:30. Total running time: 20min 2s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        6           338.216    1.38184       0.4824 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1     

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000006)


(func pid=9959) [8,  2000] loss: 1.316
(func pid=9959) [8,  4000] loss: 0.659

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:08:31. Total running time: 21min 3s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        7           394.754    1.3859        0.4921 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000007)


(func pid=9959) [9,  2000] loss: 1.282
(func pid=9959) [9,  4000] loss: 0.640

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:09:31. Total running time: 22min 3s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        8           451.835    1.30289       0.5224 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000008)


(func pid=9959) [10,  2000] loss: 1.251
(func pid=9959) [10,  4000] loss: 0.629
(func pid=9959) [10,  6000] loss: 0.420

Trial status: 7 TERMINATED | 1 RUNNING | 2 PENDING
Current time: 2024-11-14 08:10:31. Total running time: 23min 3s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00007   RUNNING         4      8   0.000425594              4        9           507.988    1.36195       0.5003 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_

(func pid=9959) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00007_7_batch_size=4,l1=4,l2=8,lr=0.0004_2024-11-14_07-47-28/checkpoint_000009)



Trial status: 8 TERMINATED | 2 PENDING
Current time: 2024-11-14 08:11:01. Total running time: 23min 33s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |
| train_cifar_b2154_00002   TERMINATED      8      4   0.00603131               4        1            58.1518   2.29459       0.1152 |
| train_ci

(func pid=13211) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=13211)   warnings.warn(


(func pid=13211) [1,  2000] loss: 2.338
(func pid=13211) [1,  4000] loss: 1.172

Trial status: 8 TERMINATED | 1 RUNNING | 1 PENDING
Current time: 2024-11-14 08:11:31. Total running time: 24min 3s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00008   RUNNING        32      2   0.035713                 2                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128  

(func pid=13211) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00008_8_batch_size=2,l1=32,l2=2,lr=0.0357_2024-11-14_07-47-28/checkpoint_000000)



Trial train_cifar_b2154_00009 started with configuration:
+--------------------------------------------------+
| Trial train_cifar_b2154_00009 config             |
+--------------------------------------------------+
| batch_size                                     2 |
| l1                                             1 |
| l2                                            64 |
| lr                                       0.00093 |
+--------------------------------------------------+
(func pid=13815) Files already downloaded and verified
(func pid=13815) Files already downloaded and verified


(func pid=13815) /usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:617: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
(func pid=13815)   warnings.warn(



Trial status: 9 TERMINATED | 1 RUNNING
Current time: 2024-11-14 08:13:01. Total running time: 25min 33s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00009   RUNNING         1     64   0.000930752              2                                                    |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |
| train_ci

(func pid=13815) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00009_9_batch_size=2,l1=1,l2=64,lr=0.0009_2024-11-14_07-47-28/checkpoint_000000)



Trial status: 9 TERMINATED | 1 RUNNING
Current time: 2024-11-14 08:14:31. Total running time: 27min 3s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00009   RUNNING         1     64   0.000930752              2        1            97.0339   1.93289       0.2203 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848              8        1            37.4279   2.28873       0.1537 |
| train_cif

(func pid=13815) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00009_9_batch_size=2,l1=1,l2=64,lr=0.0009_2024-11-14_07-47-28/checkpoint_000001)


(func pid=13815) [3,  2000] loss: 1.856
(func pid=13815) [3,  4000] loss: 0.915

Trial status: 9 TERMINATED | 1 RUNNING
Current time: 2024-11-14 08:16:31. Total running time: 29min 3s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00009   RUNNING         1     64   0.000930752              2        2           194.547    1.84551       0.2577 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.000232848

(func pid=13815) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00009_9_batch_size=2,l1=1,l2=64,lr=0.0009_2024-11-14_07-47-28/checkpoint_000002)


(func pid=13815) [4,  2000] loss: 1.852
(func pid=13815) [4,  4000] loss: 0.903

Trial status: 9 TERMINATED | 1 RUNNING
Current time: 2024-11-14 08:18:02. Total running time: 30min 33s
Logical resource usage: 2.0/2 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:T4)
+------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                status         l1     l2            lr     batch_size     iter     total time (s)      loss     accuracy |
+------------------------------------------------------------------------------------------------------------------------------------+
| train_cifar_b2154_00009   RUNNING         1     64   0.000930752              2        3           290.298    1.82112       0.2594 |
| train_cifar_b2154_00000   TERMINATED    128     32   0.000342318             16       10           268.076    1.32249       0.5215 |
| train_cifar_b2154_00001   TERMINATED     32    128   0.00023284

2024-11-14 08:19:19,937	INFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/root/ray_results/train_cifar_2024-11-14_07-47-28' in 0.0090s.



Trial train_cifar_b2154_00009 finished iteration 4 at 2024-11-14 08:19:19. Total running time: 31min 51s
+------------------------------------------------------------+
| Trial train_cifar_b2154_00009 result                       |
+------------------------------------------------------------+
| checkpoint_dir_name                      checkpoint_000003 |
| time_this_iter_s                                  96.35126 |
| time_total_s                                     386.64942 |
| training_iteration                                       4 |
| accuracy                                            0.2807 |
| loss                                               1.80892 |
+------------------------------------------------------------+
Trial train_cifar_b2154_00009 saved a checkpoint for iteration 4 at: (local)/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00009_9_batch_size=2,l1=1,l2=64,lr=0.0009_2024-11-14_07-47-28/checkpoint_000003

Trial train_cifar_b2154_00009 completed

(func pid=13815) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/root/ray_results/train_cifar_2024-11-14_07-47-28/train_cifar_b2154_00009_9_batch_size=2,l1=1,l2=64,lr=0.0009_2024-11-14_07-47-28/checkpoint_000003)


Files already downloaded and verified
Files already downloaded and verified
Best trial test set accuracy: 0.5428


If you run the code, an example output could look like this:

``` {.sh}
Number of trials: 10/10 (10 TERMINATED)
+-----+--------------+------+------+-------------+--------+---------+------------+
| ... |   batch_size |   l1 |   l2 |          lr |   iter |    loss |   accuracy |
|-----+--------------+------+------+-------------+--------+---------+------------|
| ... |            2 |    1 |  256 | 0.000668163 |      1 | 2.31479 |     0.0977 |
| ... |            4 |   64 |    8 | 0.0331514   |      1 | 2.31605 |     0.0983 |
| ... |            4 |    2 |    1 | 0.000150295 |      1 | 2.30755 |     0.1023 |
| ... |           16 |   32 |   32 | 0.0128248   |     10 | 1.66912 |     0.4391 |
| ... |            4 |    8 |  128 | 0.00464561  |      2 | 1.7316  |     0.3463 |
| ... |            8 |  256 |    8 | 0.00031556  |      1 | 2.19409 |     0.1736 |
| ... |            4 |   16 |  256 | 0.00574329  |      2 | 1.85679 |     0.3368 |
| ... |            8 |    2 |    2 | 0.00325652  |      1 | 2.30272 |     0.0984 |
| ... |            2 |    2 |    2 | 0.000342987 |      2 | 1.76044 |     0.292  |
| ... |            4 |   64 |   32 | 0.003734    |      8 | 1.53101 |     0.4761 |
+-----+--------------+------+------+-------------+--------+---------+------------+

Best trial config: {'l1': 64, 'l2': 32, 'lr': 0.0037339984519545164, 'batch_size': 4}
Best trial final validation loss: 1.5310075663924216
Best trial final validation accuracy: 0.4761
Best trial test set accuracy: 0.4737
```

Most trials have been stopped early in order to avoid wasting resources.
The best performing trial achieved a validation accuracy of about 47%,
which could be confirmed on the test set.

So that\'s it! You can now tune the parameters of your PyTorch models.
